firefox_puppeteer/api/software_update.py

```diff
@@ -0,0 +1,544 @@
+        info = update_data['updates'][update_data['update']['index']]
+
+        # The upgraded version should be identical with the version given by
+        # the update and we shouldn't have run a downgrade
+        check = self.marionette.execute_script("""
+            Cu.import("resource://gre/modules/Services.jsm");
+
+            return  Services.vc.compare(arguments[0], arguments[1]);
+        """, script_args=[info['build_post']['version'], info['build_pre']['version']])
+
+        assert check >= 0, 'The version of the upgraded build is higher or equal'
+
+        # If there was an update, the post build id should be identical with the patch
+        if info['patch']['buildid']:
+            assert info['build_post']['buildid'] == info['patch']['buildid'], \
+                'The post buildid is equal to the buildid of the update'
+
+        # If a target build id has been given, check if it matches the updated build
+        info['target_buildid'] = update_data['update'].get('target_buildid')
+        if info['target_buildid']:
+            assert info['build_post']['buildid'] == info['target_buildid'], \
+                'Post buildid matches target buildid of the update patch'
+
+        # An upgrade should not change the builds locale
+        assert info['build_post']['locale'] == info['build_pre']['locale'],\
+            'The locale of the updated build has not been changed'
+
+        # Check that no application-wide add-ons have been disabled
+        assert info['build_post']['disabled_addons'] == info['build_pre']['disabled_addons'],\
+            'No application-wide add-ons have been disabled by the update'
+
+    @property
+    def build_info(self):
+        """Return information of the current build version
+
+        :returns: A dictionary of build information
+        """
+
+        return {
+            'buildid': self.app_info.appBuildID,
+            'channel': self.update_channel.channel,
+            'disabled_addons': self.prefs.get_pref('PREF_DISABLED_ADDONS'),
+            'locale': self.app_info.locale,
+            'mar_channels': self.mar_channels.channels,
+            'url_aus': self.update_url(True),
+            'user_agent': self.app_info.user_agent,
+            'version': self.app_info.version
+        }
+
+    def force_fallback(self):
+        """Update the update.status file and set the status to 'failed:6'"""
+        with open(os.path.join(self.staging_directory, 'update.status'), 'w') as f:
+            f.write('failed: 6\n')
+
+    def get_update_at(self, update_index):
+        """Use nsIUpdateManager.getUpdateAt() to return an update
+
+        :returns: nsIUpdate object
+        """
+
+        return self.marionette.execute_script("""
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            let update_index = arguments[0];
+            return ums.getUpdateAt(update_index);
+
+        """, script_args=[update_index])
```

Comment:
It turns out I never need to call this `get_update_at` anyway. I think I added it while testing some things, so I will just remove the entire method.